homie/src/lib.rs

```diff
@@ -317,13 +350,39 @@ impl HomieDevice {
         self.publisher.publish_retained("$nodes", node_ids).await
     }
 
+    async fn set_state(&mut self, state: State) -> Result<(), SendError<Request>> {
+        self.state = state;
+        self.publisher.publish_retained("$state", self.state).await
+    }
+
+    /// Update the [state](https://homieiot.github.io/specification/#device-lifecycle) of the Homie
+    /// device to 'ready'. This should be called once it is ready to begin normal operation, or to
+    /// return to normal operation after calling `sleep()` or `alert()`.
     pub async fn ready(&mut self) -> Result<(), SendError<Request>> {
-        assert_eq!(self.state, State::Init);
-        self.state = State::Ready;
+        assert!(&[State::Init, State::Sleeping, State::Alert].contains(&self.state));
+        self.set_state(State::Ready).await
+    }
 
-        self.publisher.publish_retained("$state", "ready").await?;
+    /// Update the [state](https://homieiot.github.io/specification/#device-lifecycle) of the Homie
+    /// device to 'sleeping'. This should be only be called after `ready()`, otherwise it will panic.
+    pub async fn sleep(&mut self) -> Result<(), SendError<Request>> {
+        assert_eq!(self.state, State::Ready);
+        self.set_state(State::Sleeping).await
+    }
 
-        Ok(())
+    /// Update the [state](https://homieiot.github.io/specification/#device-lifecycle) of the Homie
+    /// device to 'alert', to indicate that something wrong is happening and manual intervention may
+    /// be required. This should be only be called after `ready()`, otherwise it will panic.
+    pub async fn alert(&mut self) -> Result<(), SendError<Request>> {
+        assert_eq!(self.state, State::Ready);
+        self.set_state(State::Alert).await
+    }
+
+    /// Disconnect cleanly from the MQTT broker, after updating the state of the Homie device to
+    // 'disconnected'.
+    pub async fn disconnect(mut self) -> Result<(), SendError<Request>> {
+        self.set_state(State::Disconnected).await?;
+        self.publisher.disconnect().await
```

Comment:
I wasn't able to test this last night, but it turns out it doesn't, it panics with unimplemented, because rumqttc doesn't actually handle disconnection. I guess I'll leave this out for now.